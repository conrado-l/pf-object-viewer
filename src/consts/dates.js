export default {
  SHORT_DATE_FORMAT: 'MM/DD/YYYY'
}
