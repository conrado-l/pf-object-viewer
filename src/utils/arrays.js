/**
 * Checks if the value is contained in the array looking by a property in the array.
 * @param {number|string} findValue
 * @param {array} searchArray
 * @param {string} propertyName
 * @returns {boolean|null}
 */
export function hasPropertyValueInArray (findValue, searchArray, propertyName) {
  if (!findValue || !searchArray || !propertyName) {
    return null
  }

  return searchArray.find(element => element[propertyName] === findValue) !== undefined
}

/**
 * Returns an array with the values that matches another array by a property.
 * @param {array} arrayA // TODO: rename param
 * @param {array} arrayB // TODO: rename param
 * @param {string} propertyName
 * @returns {array}
 */
export function arrayIncludesArrayByProperty (arrayA, arrayB, propertyName) {
  if (!arrayA || !arrayB || !propertyName) {
    return null
  }

  return arrayA.filter(rawValue =>
    arrayB.find(element => element[propertyName] === rawValue) !== undefined
  )
}
