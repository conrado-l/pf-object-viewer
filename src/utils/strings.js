/**
 * Checks if the string has comma delimitators.
 * @param {string} string String to be parsed
 */
export function hasCommaDelimitator (string) {
  if (typeof string === 'string' || string instanceof String) {
    return string.indexOf(',') !== -1
  } else {
    return false
  }
}
