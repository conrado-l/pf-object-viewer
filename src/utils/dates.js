import moment from 'moment'
import dates from '@/consts/dates'
/**
 * Generates a short date format from unix time
 * @param {number} unixtime
 */
export function unixtimeToShortDate (unixtime) {
  return moment.unix(unixtime).format(dates.SHORT_DATE_FORMAT)
}
