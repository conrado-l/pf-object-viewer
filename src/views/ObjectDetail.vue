<template>
  <v-layout>
    <v-flex xs12 md4 offset-md4>
      <card-detail
              title="Object detail"
              :items="objectDetail"
              :isLoading="isLoading"
              :error="error"
              error-message="An error has occurred with the object detail."
      >
      </card-detail>
      <v-flex v-if="error" text-xs-center>
        <v-btn color="info" @click="fetchObjectDetail">Retry</v-btn>
      </v-flex>
    </v-flex>
  </v-layout>
</template>

<script>
/**
 * Fetches an object's detail data by ID and renders it in a card detail.
 */
import { mapGetters } from 'vuex'
import loaders from '@/consts/loaders'
import CardDetail from '@/components/CardDetail'
import { unixtimeToShortDate } from '@/utils/dates'

export default {
  name: 'ObjectDetail',
  components: { CardDetail },
  props: {
    // Passed from the router (props: true)
    // Other approach would be to create a variable "id" in data () and assign this.$route.params.id to it
    id: {
      type: String,
      required: true,
      default: ''
    }
  },
  data () {
    return {
      error: false
    }
  },
  methods: {
    /**
     * Fetches the object detail with the associated ID.
     */
    fetchObjectDetail () {
      this.$store.dispatch('objectDetail/fetchObject', this.id)
        .then(() => {
          this.error = false // Set it to false in case it failed and the user clicks the Retry button
        })
        .catch(() => {
          this.error = true
        })
    }
  },
  created () {
    // No ID was found in the URL
    if (!this.id) {
      this.error = true
      return
    }

    this.fetchObjectDetail()
  },
  computed: {
    /**
     * Generates the object detail with the associated description label.
     * @returns {object}
     * Consideration: the server could also send the label but i18n should be server-side then.
     */
    objectDetail () {
      if (!this.getObject) {
        return []
      }

      return [
        { description: 'ID', value: this.getObject.id },
        { description: 'Name', value: this.getObject.name },
        { description: 'Description', value: this.getObject.description },
        { description: 'Type', value: this.getObject.type },
        { description: 'Available', value: this.getObject.available ? 'Yes' : 'No' },
        { description: 'Creation Date', value: unixtimeToShortDate(this.getObject.creation_date) }
      ]
    },
    /**
     * Indicates if there is a fetching operation at the moment.
     * @returns {boolean}
     */
    isLoading () {
      return this.$wait.is(loaders.objectDetail.FETCH_OBJECT_DETAIL)
    },
    ...mapGetters('objectDetail', [
      'getObject'
    ])
  },
  /**
   * Resets the store when the component is about the be destroyed to avoid showing old data when fetching new one.
   */
  beforeDestroy () {
    this.$store.commit('objectDetail/reset')
  }
}
</script>

<style lang="scss" scoped>

</style>
