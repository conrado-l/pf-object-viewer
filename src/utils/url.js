/**
 * Generates a URL query based on the object properties
 * @param {object} paramsObject
 * @example {value: 1, name: 'George'} --> '?value=1&name=George'
 * @return {string}
 */
import { hasCommaDelimitator } from '@/utils/strings'

export function generateURLQueryFromObject (paramsObject) {
  const query = new URLSearchParams(paramsObject).toString()
  return `?${query}`
}

/**
 * Returns an object in which every property with multiple values separated by a comma, become a split array.
 * @param {object} query
 * @example {sort: description,name} = {sort: ['description','name']}
 * @returns {object}
 */
export function parseRouteQuery (query) {
  if (!query || !Object.keys(query).length) {
    return {}
  }

  let parsedQuery = {}

  Object.keys(query).forEach((key) => { // Object.Keys already checks for hasOwnProperty
    const value = query[key]

    if (value) {
      parsedQuery[key] = hasCommaDelimitator(value) ? value.split(',') : value // TODO: case "search=text,"
    }
  })

  return parsedQuery
}
