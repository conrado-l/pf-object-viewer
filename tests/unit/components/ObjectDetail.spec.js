import Vue from 'vue'
import { shallowMount } from '@vue/test-utils'
import ObjectDetail from '@/views/ObjectDetail'
import Vuetify from 'vuetify'
import { Store } from 'vuex-mock-store'
import { unixtimeToShortDate } from '@/utils/dates'

describe('ObjectDetail.vue', () => {
  const mockupObject = {
    available: false,
    creation_date: 1553276681,
    description: 'An amazing blue shirt',
    id: '125',
    name: 'Shirt',
    type: 'clothing'
  }

  const store = new Store({
    state: {
      objectDetail: {
        object: mockupObject
      }
    },
    getters: {
      'objectDetail/getObject': mockupObject
    }
  })

  const factoryShallowMount = (props, loading = false) => {
    return shallowMount(ObjectDetail, {
      propsData: {
        id: mockupObject.id,
        ...props
      },
      mocks: {
        $store: store,
        $wait: {
          is: () => {
            return loading
          }
        }
      }
    })
  }

  beforeEach(() => {
    Vue.use(Vuetify) // Vuetify issue with createLocalVue: https://github.com/vuetifyjs/vuetify/issues/4964
    store.dispatch = jest.fn().mockResolvedValue('data') // dispatch has to return a promise
  })

  afterEach(() => {
    store.reset()
  })

  it('should render the component correctly', () => {
    const wrapper = factoryShallowMount()

    expect(wrapper.element).toMatchSnapshot()
  })

  it('should set error to true if the object id route param does not exists', () => {
    const wrapper = factoryShallowMount({ id: '' })

    expect(wrapper.vm.error).toBe(true)
  })

  it('should set error to false if the object id route param exists', () => {
    const wrapper = factoryShallowMount()

    expect(wrapper.vm.error).toBe(false)
  })

  it('should dispatch fetchObject action with the object id as payload', () => {
    factoryShallowMount()

    expect(store.dispatch).toHaveBeenCalledWith('objectDetail/fetchObject', mockupObject.id)
  })

  it('should return an empty array if getObject getter is not set yet from the objectDetail computed property', () => {
    const wrapper = factoryShallowMount()
    store.getters['objectDetail/getObject'] = null

    expect(wrapper.vm.objectDetail).toEqual([])
  })

  it('should return an array with the correct object values and descriptions from the objectDetail computed property', () => {
    const wrapper = factoryShallowMount()
    const getObjectGetter = store.getters['objectDetail/getObject']

    expect(wrapper.vm.objectDetail).toEqual([
      { description: 'ID', value: getObjectGetter.id },
      { description: 'Name', value: getObjectGetter.name },
      { description: 'Description', value: getObjectGetter.description },
      { description: 'Type', value: getObjectGetter.type },
      { description: 'Available', value: 'No' },
      { description: 'Creation Date', value: unixtimeToShortDate(getObjectGetter.creation_date) }
    ])
  })

  it('should set the isLoading computed property to true on loading', () => {
    const wrapper = factoryShallowMount(null, true)

    expect(wrapper.vm.isLoading).toBe(true)
  })

  it('should dispatch fetchObject action and set error to true if the action fails', () => {
    // store.dispatch = jest.fn().mockRejectedValue('data')
    //
    // const wrapper = factoryShallowMount()
    // wrapper.vm.$nextTick(() => {
    //   expect(wrapper.vm.error).toBe(true)
    // })
  })
})
