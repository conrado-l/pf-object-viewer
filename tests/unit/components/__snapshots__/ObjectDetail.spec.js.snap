// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObjectDetail.vue should render the component correctly 1`] = `
<v-layout-stub
  tag="div"
>
  <v-flex-stub
    md4=""
    offset-md4=""
    tag="div"
    xs12=""
  >
    <card-detail-stub
      errormessage="An error has occurred with the object detail."
      items="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
      title="Object detail"
    />
     
    <!---->
  </v-flex-stub>
</v-layout-stub>
`;
